src/main/resources/db/migration_mariadb/V4__foreign_keys.sql

```diff
@@ -0,0 +1,423 @@
+
+drop index at_state_id_next_height_height_idx on at_state;
+create index at_state_id_next_height_height_idx
+	on at_state (at_id asc, next_height asc, height desc);
+
+alter table at_state
+	add constraint at_state_at_id_fk
+		foreign key (at_id) references at (id)
+			on delete cascade;
+
+-- bid_order
+
+drop index bid_order_account_id_idx on bid_order;
+create index bid_order_account_id_idx
+	on bid_order (account_id asc, height desc);
+
+drop index bid_order_creation_idx on bid_order;
+create index bid_order_creation_idx
+	on bid_order (creation_height desc);
+
+drop index bid_order_id_height_idx on bid_order;
+create unique index bid_order_id_height_idx
+	on bid_order (id asc, height desc);
+
+alter table bid_order
+	add constraint bid_order_account_id_fk
+		foreign key (account_id) references account (id)
+			on delete cascade;
+
+alter table bid_order
+	add constraint bid_order_asset_id_fk
+		foreign key (asset_id) references asset (id)
+			on delete cascade;
+
+-- block
+-- no generator_id foreign key because account may be not yet created
+
+drop index block_timestamp_idx on block;
+create unique index block_timestamp_idx
+	on block (timestamp desc);
+
+-- escrow
+
+drop index escrow_deadline_height_idx on escrow;
+create index escrow_deadline_height_idx
+	on escrow (deadline asc, height desc);
+
+drop index escrow_id_height_idx on escrow;
+create unique index escrow_id_height_idx
+	on escrow (id asc, height desc);
+
+drop index escrow_recipient_id_height_idx on escrow;
+create index escrow_recipient_id_height_idx
+	on escrow (recipient_id asc, height desc);
+
+drop index escrow_sender_id_height_idx on escrow;
+create index escrow_sender_id_height_idx
+	on escrow (sender_id asc, height desc);
+
+alter table escrow
+	add constraint escrow_account_id_fk
+		foreign key (sender_id) references account (id)
+			on delete cascade;
+
+alter table escrow
+	add constraint escrow_account_id_fk_2
+		foreign key (recipient_id) references account (id)
+			on delete cascade;
+
+-- escrow_decision
+
+drop index escrow_decision_account_id_height_idx on escrow_decision;
+create index escrow_decision_account_id_height_idx
+	on escrow_decision (account_id asc, height desc);
+
+drop index escrow_decision_escrow_id_account_id_height_idx on escrow_decision;
+create unique index escrow_decision_escrow_id_account_id_height_idx
+	on escrow_decision (escrow_id asc, account_id asc, height desc);
+
+drop index escrow_decision_escrow_id_height_idx on escrow_decision;
+create index escrow_decision_escrow_id_height_idx
+	on escrow_decision (escrow_id asc, height desc);
+
+alter table escrow_decision
+	add constraint escrow_decision_account_id_fk
+		foreign key (account_id) references account (id)
+			on delete cascade;
+
+alter table escrow_decision
+	add constraint escrow_decision_escrow_id_fk
+		foreign key (escrow_id) references escrow (id)
+			on delete cascade;
+
+-- goods
+
+drop index goods_id_height_idx on goods;
+create unique index goods_id_height_idx
```

Comment:
Will lines like this cause the DB to completely rebuild the index or just build it for future insertions?

Reply:
Completely rebuild, we drop index and create it again, it's a heavy operation and give effect(some perfomance improving on select) only if backend MySQL/Oracle, not MariadDB - but they are currently fully compatible.
https://dev.mysql.com/doc/refman/8.0/en/descending-indexes.html https://jira.mariadb.org/browse/MDEV-13756

If MariaDB will add desc index ordering, it will give effect only  after reinstallation from scratch.

It will be nice if first migration was contain desc index ordering(same as h2), and now I don't mind remove this lines if you say.